engine/src/main/java/org/hibernate/validator/constraintvalidation/HibernateValidatorFactoryObserver.java

```diff
@@ -0,0 +1,43 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ * Copyright Red Hat Inc. and Hibernate Authors
+ */
+package org.hibernate.validator.constraintvalidation;
+
+import org.hibernate.validator.HibernateValidatorFactory;
+import org.hibernate.validator.Incubating;
+
+/**
+ * Allows observing the basic validator factory lifecycle events.
+ * <p>
+ * Observers are required to handle exceptions internally and must <b>not</b> propagate them.
+ *
+ * @see org.hibernate.validator.HibernateValidatorConfiguration#addHibernateValidatorFactoryObserver(HibernateValidatorFactoryObserver)
+ * @since 9.1.0
+ */
+@Incubating
+public interface HibernateValidatorFactoryObserver {
+	/**
+	 * A callback invoked upon the successful creation of a factory.
+	 *
+	 * @param factory The fully initialized factory.
+	 */
+	default void factoryCreated(HibernateValidatorFactory factory) {
+	}
+
+	/**
+	 * A callback invoked just before the factory is shut down.
+	 *
+	 * @param factory The factory that is about to be closed.
+	 */
+	default void factoryClosing(HibernateValidatorFactory factory) {
+	}
+
+	/**
+	 * A callback invoked after the factory has been successfully shut down.
+	 *
+	 * @param factory The closed factory.
+	 */
+	default void factoryClosed(HibernateValidatorFactory factory) {
+	}
```

Comment:
and since clearing the cache for the pattern "shared service" explicitly felt more like a "hack" I've added this observer to react to HV factory lifecycle events and used it to clear the cache instead...

Reply:
Hmm, I'll have to see how it is done for the predefined scope... I don't remember if the init context there is "short lived" or if it stays available for the whole lifetime of the factory... it should stay available for the regular case, as for it, the metadata is built lazily on the first validation attempt...